sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchedStreamingWrite.java

```diff
@@ -0,0 +1,402 @@
+    private transient Map<String, List<FailsafeValueInSingleWindow<TableRow, TableRow>>> tableRows;
+
+    /** The list of unique ids for each BigQuery table row. */
+    private transient Map<String, List<String>> uniqueIdsForTableRows;
+
+    @Setup
+    public void setup() {
+      // record latency upto 60 seconds in the resolution of 20ms
+      histogram = Histogram.linear(0, 20, 3000);
+      lastReportedSystemClockMillis = System.currentTimeMillis();
+    }
+
+    @Teardown
+    public void teardown() {
+      if (histogram.getTotalCount() > 0) {
+        logPercentiles();
+        histogram.clear();
+      }
+    }
+
+    /** Prepares a target BigQuery table. */
+    @StartBundle
+    public void startBundle() {
+      tableRows = new HashMap<>();
+      uniqueIdsForTableRows = new HashMap<>();
+    }
+
+    /** Accumulates the input into JsonTableRows and uniqueIdsForTableRows. */
+    @ProcessElement
+    public void processElement(
+        @Element KV<String, TableRowInfo<ElementT>> element,
+        @Timestamp Instant timestamp,
+        BoundedWindow window,
+        PaneInfo pane) {
+      String tableSpec = element.getKey();
+      List<FailsafeValueInSingleWindow<TableRow, TableRow>> rows =
+          BigQueryHelpers.getOrCreateMapListValue(tableRows, tableSpec);
+      List<String> uniqueIds =
+          BigQueryHelpers.getOrCreateMapListValue(uniqueIdsForTableRows, tableSpec);
+
+      TableRow tableRow = toTableRow.apply(element.getValue().tableRow);
+      TableRow failsafeTableRow = toFailsafeTableRow.apply(element.getValue().tableRow);
+      rows.add(FailsafeValueInSingleWindow.of(tableRow, timestamp, window, pane, failsafeTableRow));
+      uniqueIds.add(element.getValue().uniqueId);
+    }
+
+    /** Writes the accumulated rows into BigQuery with streaming API. */
+    @FinishBundle
+    public void finishBundle(FinishBundleContext context) throws Exception {
+      List<ValueInSingleWindow<ErrorT>> failedInserts = Lists.newArrayList();
+      BigQueryOptions options = context.getPipelineOptions().as(BigQueryOptions.class);
+      for (Map.Entry<String, List<FailsafeValueInSingleWindow<TableRow, TableRow>>> entry :
+          tableRows.entrySet()) {
+        TableReference tableReference = BigQueryHelpers.parseTableSpec(entry.getKey());
+        flushRows(
+            tableReference,
+            entry.getValue(),
+            uniqueIdsForTableRows.get(entry.getKey()),
+            options,
+            failedInserts);
+      }
+      tableRows.clear();
+      uniqueIdsForTableRows.clear();
+
+      for (ValueInSingleWindow<ErrorT> row : failedInserts) {
+        context.output(failedOutputTag, row.getValue(), row.getTimestamp(), row.getWindow());
+      }
+
+      updateAndLogHistogram(options);
+    }
+  }
+
+  private class ViaStateful
+      extends PTransform<PCollection<KV<String, TableRowInfo<ElementT>>>, PCollection<ErrorT>> {
+    private final Duration BATCH_MAX_BUFFERING_DURATION = Duration.millis(200);
```

Comment:
Not sure about this limit. What would be a proper value? Should we make it configurable?

Reply:
to be hones, I am not sure what's a good duration either. I think this is acceptable for now, until we find out more. Thoughts?

Reply:
Yeah sounds reasonable to proceed with this for now.